src/core/processing/qgsprocessingparameters.h

```diff
@@ -1566,6 +1582,37 @@ class CORE_EXPORT QgsProcessingParameterPoint : public QgsProcessingParameterDef
 
 };
 
+/**
+ * \class QgsProcessingParameterGeometry
+ * \ingroup core
+ * A geometry parameter for processing algorithms.
+  * \since QGIS 3.16
+ */
+class CORE_EXPORT QgsProcessingParameterGeometry : public QgsProcessingParameterDefinition
+{
+  public:
+
+    /**
+     * Constructor for QgsProcessingParameterGeometry.
+     */
+    QgsProcessingParameterGeometry( const QString &name, const QString &description = QString(), const QVariant &defaultValue = QVariant(), bool optional = false );
```

Comment:
I think this constructor should also accept a list of permissible geometry types, e..g to allow parameters which only accept line values or line OR polygon values, etc

Reply:
If we use QgsWkbTypes::GeometryType there is no semantic for 'any geometries'.

Reply:
I'm in favor to postpone this filter atm,   In the situations where  ones would want to restrict the geometry type, it may always be transformed by the caller to fit the needs. I'm not sure there is a real value added here. If this is really needed at some point this could be implemented later. Let the usage tel us.

Reply:
Done

Note that ATM, I had to decorate QgsReferencedGeometry with SIP_SKIP because it breaks sip code generation (see notes below).

Reply:
> Done
> 
> Note that ATM, I had to decorate QgsReferencedGeometry with SIP_SKIP because it breaks sip code generation (see notes below).


I have found that:
- If I skip another classe (ex QgsReferencedPointXY) instead QgsReferencedGeometry then the compilation succeed
- The compilation fail as soon as we sipify any extra class (whatever the class is, even a simple classe declaration with empty body).

Is there a limitation in the number of class  that sip can handel of do we reach the limit of a buffer ?

Reply:
> If we use QgsWkbTypes::GeometryType there is no semantic for 'any geometry'.

QgsWkbTypes::UnknownGeometry can be used to indicate this.

But you'd need to accept a list here anyway (to allow for e.g. lines or polygons but not points, so potentially you could also just populate that list with point/line/polygon to indicate any types are allowed.

> I'm in favor to postpone this filter atm, In the situations where ones would want to restrict the geometry type, it may always be transformed to fit the needs. I'm not sure there is a real value added here. If this is really needed at some point this could be implemented later. Let the usage tel us.

My gut feeling is that the times when an algorithm CAN work with any geometry type will be much much less than the times when a specific type is required. Use cases I'm thinking of for the new parameter type are things like a "profile tool" algorithm where uses select a line geometry to generate the profile along -- in this case, you'd definitely have to prevent points or polygon inputs.

Reply:
Done, with tests

- You may pass a list of QgsWbkTypes::GeometryType as parameter: `checkValueIsAcceptable`  will check the type of geometry represented by the passed values against the values contained in the list - 
If the list is empty or the list contains `QgsWkbTypes::UnknownGeometry ` the test always succeed, otherewise the test succeed if the list contains the tested type.
- Variants map et python string are tested against this parameter.